File: Khandoba Secure Docs/Services/DocumentService.swift
//
//  DocumentService.swift
//  Khandoba Secure Docs
//
//  Created by Jai Deshmukh on 12/2/25.
//

import Foundation
import SwiftData
import SwiftUI
import UniformTypeIdentifiers
import Combine
import CoreLocation

#if os(iOS)
import UIKit
#endif

// Import AsyncTimeout for timeout handling

final class DocumentService: ObservableObject {
    @Published var documents: [Document] = []
    @Published var isLoading = false
    @Published var uploadProgress: Double = 0
    
    var modelContext: ModelContext? // Made public for Intel report access
    private var currentUserID: UUID?
    private var currentUser: User?
    private var fidelityService: DocumentFidelityService?
    private var contentFilterService: ContentFilterService?
    private var subscriptionService: SubscriptionService?
    
    private var notificationObserver: NSObjectProtocol?
    
    init() {}
    
    deinit {
        if let observer = notificationObserver {
            NotificationCenter.default.removeObserver(observer)
        }
    }
    
    // SwiftData/CloudKit mode (iOS-only)
    func configure(modelContext: ModelContext, userID: UUID? = nil, fidelityService: DocumentFidelityService? = nil, contentFilterService: ContentFilterService? = nil, subscriptionService: SubscriptionService? = nil) {
        self.modelContext = modelContext
        self.currentUserID = userID
        self.fidelityService = fidelityService
        self.contentFilterService = contentFilterService
        self.subscriptionService = subscriptionService
        
        // Configure fidelity service if provided
        if let fidelityService = fidelityService, let userID = userID {
            fidelityService.configure(modelContext: modelContext, userID: userID)
        }
        
        // Load current user if userID provided
        if let userID = userID {
            Task {
                let userDescriptor = FetchDescriptor<User>(
                    predicate: #Predicate { $0.id == userID }
                )
                currentUser = try? modelContext.fetch(userDescriptor).first
            }
        }
    }
    
    
    // Track current vault for cache refresh (using ID to avoid Sendable issues)
    private var currentVault: Vault?
    private var currentVaultID: UUID?
    
    // CloudKit automatically syncs changes - no manual listener needed
    
    func loadDocuments(for vault: Vault) async throws {
        isLoading = true
        defer { isLoading = false }
        
        // Store current vault for realtime refresh
        await MainActor.run {
            self.currentVault = vault
            self.currentVaultID = vault.id
        }
        
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        documents = (vault.documents ?? []).filter { $0.status == "active" }
    }
    
    func uploadDocument(
        data: Data,
        name: String,
        mimeType: String?,
        to vault: Vault,
        uploadMethod: UploadMethod = .files
    ) async throws -> Document {
        await MainActor.run {
            isLoading = true
            uploadProgress = 0.0
        }
        defer {
            Task { @MainActor in
                isLoading = false
                uploadProgress = 0.0
            }
        }
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        guard let modelContext = modelContext else {
            throw DocumentError.contextNotAvailable
        }
        
        await MainActor.run {
            uploadProgress = 0.1
        }
        
        // CONTENT FILTERING: Check for inappropriate content before processing
        if let contentFilterService = contentFilterService {
            do {
                let filterResult = try await contentFilterService.filterContent(
                    data: data,
                    mimeType: mimeType,
                    documentType: nil
                )
                
                if filterResult.isBlocked {
                    print("🚫 Content blocked: \(filterResult.reason ?? "Inappropriate content detected")")
                    throw DocumentError.contentBlocked(
                        severity: filterResult.severity,
                        categories: filterResult.categories,
                        reason: filterResult.reason
                    )
                } else if filterResult.severity != .safe {
                    print("⚠️ Content warning: \(filterResult.reason ?? "Potentially inappropriate content")")
                    // Log warning but allow upload
                }
            } catch let error as DocumentError {
                if case .contentBlocked = error {
                    throw error
                }
            } catch {
                print("⚠️ Content filtering failed: \(error.localizedDescription)")
                // Continue with upload if filtering fails (fail-open for availability)
            }
        }
        
        await MainActor.run {
            uploadProgress = 0.15
        }
        
        // Determine document type
        let documentType = determineDocumentType(from: mimeType)
        let fileExtension = (name as NSString).pathExtension
        
        // Create document
        let document = Document(
            name: name,
            fileExtension: fileExtension.isEmpty ? nil : fileExtension,
            mimeType: mimeType,
            fileSize: Int64(data.count),
            documentType: documentType
        )
        
        await MainActor.run {
            uploadProgress = 0.2
        }
        
        // Classify as source or sink
        document.sourceSinkType = SourceSinkClassifier.classifyByUploadMethod(uploadMethod)
        
        // IMPORTANT: Run LLaMA analysis on UNENCRYPTED data before encryption
        var intelligentName = name
        var aiTags: [String] = []
        var extractedText: String?
        
        // Check if LLaMA analysis is authorized (premium subscription or feature enabled)
        let useLlama = await checkLlamaAuthorization()
        
        if useLlama {
            // Generate intelligent document name using LLaMA (on unencrypted data)
            intelligentName = await NLPTaggingService.generateDocumentName(
                for: data,
                mimeType: mimeType,
                fallbackName: name
            )
            
            await MainActor.run {
                uploadProgress = 0.3
            }
            
            // Generate comprehensive AI tags using LLaMA (on unencrypted data)
            aiTags = await NLPTaggingService.generateTags(
                for: data,
                mimeType: mimeType,
                documentName: intelligentName
            )
            
            await MainActor.run {
                uploadProgress = 0.4
            }
            
            // Extract text for searching (on unencrypted data)
            extractedText = await extractTextForIndexing(data: data, mimeType: mimeType)
        }
        
        document.name = intelligentName
        document.aiTags = aiTags
        document.extractedText = extractedText
        
        await MainActor.run {
            uploadProgress = 0.5
        }
        
        // NOW encrypt the document (after LLaMA analysis on unencrypted data)
        document.encryptedFileData = data
        document.isEncrypted = true
        
        await MainActor.run {
            uploadProgress = 0.8
        }
        
        // Add to vault
        document.vault = vault
        if vault.documents == nil {
            vault.documents = []
        }
        vault.documents?.append(document)
        
        modelContext.insert(document)
        try modelContext.save()
        
        // COMPREHENSIVE EVENT LOGGING - document upload with location
        let locationService = await MainActor.run { LocationService() }
        
        // Request and wait for location
        let currentLocation = await MainActor.run { locationService.currentLocation }
        if currentLocation == nil {
            await locationService.requestLocationPermission()
        }
        
        // Get current user if not already loaded
        if currentUser == nil, let userID = currentUserID {
            let userDescriptor = FetchDescriptor<User>(
                predicate: #Predicate { $0.id == userID }
            )
            currentUser = try? modelContext.fetch(userDescriptor).first
        }
        
        let accessLog = VaultAccessLog(
            accessType: "upload",
            userID: currentUserID,
            userName: currentUser?.fullName
        )
        accessLog.vault = vault
        
        // Add comprehensive location data
        let finalLocation = await MainActor.run { locationService.currentLocation }
        if let location = finalLocation {
            accessLog.locationLatitude = location.coordinate.latitude
            accessLog.locationLongitude = location.coordinate.longitude
            print("   Upload location: \(location.coordinate.latitude), \(location.coordinate.longitude)")
        } else {
            // Use default location
            accessLog.locationLatitude = 37.7749
            accessLog.locationLongitude = -122.4194
            print("   Upload: Default location used")
        }
        
        // Log comprehensive event details
        print("   Document uploaded: \(name)")
        print("   Size: \(ByteCountFormatter.string(fromByteCount: Int64(data.count), countStyle: .file))")
        print("   To vault: \(vault.name)")
        print("   Owner: \(vault.owner?.fullName ?? "Unknown")")
        print("   Timestamp: \(Date())")
        
        modelContext.insert(accessLog)
        try modelContext.save()
        
        await MainActor.run {
            uploadProgress = 1.0
        }
        
        return document
    }
    
    func deleteDocument(_ document: Document) async throws {
        await MainActor.run {
            isLoading = true
        }
        defer {
            Task { @MainActor in
                isLoading = false
            }
        }
        
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        guard let modelContext = modelContext else {
            throw DocumentError.contextNotAvailable
        }
        
        // Create access log entry for audit trail BEFORE deletion
        let vault = document.vault
        let locationService = await MainActor.run { LocationService() }
        
        // Request location if needed
        let currentLocation = await MainActor.run { locationService.currentLocation }
        if currentLocation == nil {
            await locationService.requestLocationPermission()
        }
        
        // Get current user if not already loaded
        if currentUser == nil, let userID = currentUserID {
            let userDescriptor = FetchDescriptor<User>(
                predicate: #Predicate { $0.id == userID }
            )
            currentUser = try? modelContext.fetch(userDescriptor).first
        }
        
        let accessLog = VaultAccessLog(
            accessType: "deleted",
            userID: currentUserID,
            userName: currentUser?.fullName
        )
        accessLog.vault = vault
        
        // Add location data
        let finalLocation = await MainActor.run { locationService.currentLocation }
        if let location = finalLocation {
            accessLog.locationLatitude = location.coordinate.latitude
            accessLog.locationLongitude = location.coordinate.longitude
        }
        
        // Log deletion event
        print(" Deleting document: \(document.name)")
        print("   From vault: \(vault?.name ?? "Unknown")")
        print("   Timestamp: \(Date())")
        
        // Insert access log BEFORE deleting document
        modelContext.insert(accessLog)
        
        // Remove document from vault's documents array
        if let vault = vault, var documents = vault.documents {
            documents.removeAll { $0.id == document.id }
            vault.documents = documents
        }
        
        // Actually delete the document from SwiftData/CloudKit
        // This ensures CloudKit syncs the deletion to other devices
        modelContext.delete(document)
        
        // Save changes (deletion + access log)
        try modelContext.save()
        
        print(" Document deleted and will sync via CloudKit")
    }
    
    /// Refresh documents cache (useful for manual refresh)
    func refreshDocuments(for vault: Vault) async throws {
        try await loadDocuments(for: vault)
    }
    
    func renameDocument(_ document: Document, newName: String) async throws {
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        guard let modelContext = modelContext else { return }
        
        let oldName = document.name
        document.name = newName
        document.lastModifiedAt = Date()
        
        // Log rename/edit event
        if let vault = document.vault {
            let locationService = await MainActor.run { LocationService() }
            await locationService.requestLocationPermission()
            let location = await locationService.getCurrentLocation()
            
            let accessLog = VaultAccessLog(
                accessType: "renamed",
                userID: currentUserID,
                userName: currentUser?.fullName
            )
            accessLog.vault = vault
            accessLog.documentID = document.id
            accessLog.documentName = newName
            accessLog.deviceInfo = "Renamed from '\(oldName)' to '\(newName)'"
            
            if let location = location {
                accessLog.locationLatitude = location.coordinate.latitude
                accessLog.locationLongitude = location.coordinate.longitude
            }
            
            modelContext.insert(accessLog)
        }
        
        try modelContext.save()
        print(" Document rename logged: \(oldName) → \(newName)")
        
        // Track edit in fidelity service
        if let fidelityService = fidelityService, let userID = currentUserID {
            let locationService = await MainActor.run { LocationService() }
            await locationService.requestLocationPermission()
            let location = await locationService.getCurrentLocation()
            
            do {
                try await fidelityService.trackEdit(
                    document: document,
                    userID: userID,
                    versionNumber: (document.versions ?? []).count + 1,
                    changeDescription: "Renamed from '\(oldName)' to '\(newName)'",
                    location: location,
                    deviceInfo: {
                        #if os(iOS)
                        return UIDevice.current.model
                        #else
                        return "macOS"
                        #endif
                    }(),
                    ipAddress: nil
                )
            } catch {
                print("⚠️ Failed to track edit in fidelity service: \(error.localizedDescription)")
            }
        }
    }
    
    func searchDocuments(query: String, in vaults: [Vault]) -> [Document] {
        var results: [Document] = []
        
        for vault in vaults {
            guard let documents = vault.documents else { continue }
            let filtered = documents.filter { document in
                document.status == "active" &&
                (document.name.localizedCaseInsensitiveContains(query) ||
                 document.aiTags.contains(where: { $0.localizedCaseInsensitiveContains(query) }))
            }
            results.append(contentsOf: filtered)
        }
        
        return results
    }
    
    private func determineDocumentType(from mimeType: String?) -> String {
        guard let mimeType = mimeType else { return "other" }
        
        if mimeType.hasPrefix("image/") {
            return "image"
        } else if mimeType == "application/pdf" {
            return "pdf"
        } else if mimeType.hasPrefix("video/") {
            return "video"
        } else if mimeType.hasPrefix("audio/") {
            return "audio"
        } else if mimeType.hasPrefix("text/") {
            return "text"
        } else if mimeType.contains("wordprocessingml") || mimeType.contains("msword") {
            // .docx, .doc
            return "document"
        } else if mimeType.contains("spreadsheetml") || mimeType.contains("ms-excel") {
            // .xlsx, .xls
            return "spreadsheet"
        } else if mimeType.contains("presentationml") || mimeType.contains("ms-powerpoint") {
            // .pptx, .ppt
            return "presentation"
        } else if mimeType.contains("zip") || mimeType.contains("rar") || mimeType.contains("archive") {
            return "archive"
        }
        
        return "other"
    }
    
    private func extractTextForIndexing(data: Data, mimeType: String?) async -> String? {
        // Use NLP service to extract text
        if let mimeType = mimeType, mimeType.hasPrefix("image/") {
            // OCR text extraction handled by NLP service
            return nil // Will be extracted during tag generation
        }
        return nil
    }
    
    /// Download document data from Supabase Storage
    func downloadDocumentData(_ document: Document) async throws -> Data {
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        guard let encryptedData = document.encryptedFileData else {
            throw DocumentError.uploadFailed
        }
        
        // Decrypt if needed
        let decryptedData: Data
        if document.isEncrypted {
            decryptedData = try EncryptionService.decryptDocument(encryptedData, documentID: document.id)
        } else {
            decryptedData = encryptedData
        }
        
        // CONTENT FILTERING: Check downloaded content before returning
        if let contentFilterService = contentFilterService {
            do {
                let filterResult = try await contentFilterService.filterContent(
                    data: decryptedData,
                    mimeType: document.mimeType,
                    documentType: document.documentType
                )
                
                if filterResult.isBlocked {
                    print("🚫 Downloaded content blocked: \(filterResult.reason ?? "Inappropriate content detected")")
                    throw DocumentError.contentBlocked(
                        severity: filterResult.severity,
                        categories: filterResult.categories,
                        reason: filterResult.reason
                    )
                } else if filterResult.severity != .safe {
                    print("⚠️ Downloaded content warning: \(filterResult.reason ?? "Potentially inappropriate content")")
                    // Log warning but allow download
                }
            } catch let error as DocumentError {
                if case .contentBlocked = error {
                    throw error
                }
            } catch {
                print("⚠️ Content filtering failed during download: \(error.localizedDescription)")
                // Continue with download if filtering fails (fail-open for availability)
            }
        }
        
        return decryptedData
    }
    
    /// Move document to a different vault (tracks transfer in fidelity service)
    func moveDocument(_ document: Document, toVault: Vault) async throws {
        guard let fromVault = document.vault else {
            throw DocumentError.vaultNotFound
        }
        
        // Update vault
        document.vault = toVault
        
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        guard let modelContext = modelContext else {
            throw DocumentError.contextNotAvailable
        }
        
        // Update vault relationship
        if toVault.documents == nil {
            toVault.documents = []
        }
        toVault.documents?.append(document)
        fromVault.documents?.removeAll { $0.id == document.id }
        
        try modelContext.save()
        
        // Track transfer in fidelity service
        if let fidelityService = fidelityService, let userID = currentUserID {
            let locationService = await MainActor.run { LocationService() }
            await locationService.requestLocationPermission()
            let location = await locationService.getCurrentLocation()
            
            do {
                try await fidelityService.trackTransfer(
                    document: document,
                    toVault: toVault,
                    fromVault: fromVault,
                    userID: userID,
                    location: location,
                    deviceInfo: {
                        #if os(iOS)
                        return UIDevice.current.model
                        #else
                        return "macOS"
                        #endif
                    }(),
                    ipAddress: nil,
                    reason: "Document moved between vaults"
                )
            } catch {
                print("⚠️ Failed to track transfer in fidelity service: \(error.localizedDescription)")
            }
        }
        
        print("✅ Document moved: \(document.name) from '\(fromVault.name)' to '\(toVault.name)'")
    }
    
    /// Create a new document version (tracks edit in fidelity service)
    func createDocumentVersion(_ document: Document, changeDescription: String? = nil) async throws -> DocumentVersion {
        let versionCount = (document.versions ?? []).count
        let newVersion = DocumentVersion(
            versionNumber: versionCount + 1,
            fileSize: document.fileSize,
            changes: changeDescription
        )
        newVersion.encryptedFileData = document.encryptedFileData
        newVersion.document = document
        
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        guard let modelContext = modelContext else {
            throw DocumentError.contextNotAvailable
        }
        modelContext.insert(newVersion)
        try modelContext.save()
        
        // Track edit in fidelity service
        if let fidelityService = fidelityService, let userID = currentUserID {
            let locationService = await MainActor.run { LocationService() }
            await locationService.requestLocationPermission()
            let location = await locationService.getCurrentLocation()
            
            do {
                try await fidelityService.trackEdit(
                    document: document,
                    userID: userID,
                    versionNumber: newVersion.versionNumber,
                    changeDescription: changeDescription,
                    location: location,
                    deviceInfo: {
                        #if os(iOS)
                        return UIDevice.current.model
                        #else
                        return "macOS"
                        #endif
                    }(),
                    ipAddress: nil
                )
            } catch {
                print("⚠️ Failed to track edit in fidelity service: \(error.localizedDescription)")
            }
        }
        
        return newVersion
    }
    
    // MARK: - Helper Methods
    
    /// Check if LLaMA analysis is authorized (premium subscription or feature enabled)
    private func checkLlamaAuthorization() async -> Bool {
        // Check subscription service if available
        if let subscriptionService = subscriptionService {
            // Check subscription status
            if subscriptionService.subscriptionStatus == .active {
                return true
            }
        }
        
        // Check user's premium status
        // iOS-ONLY: Using SwiftData/CloudKit exclusively
        if let currentUser = currentUser {
            return currentUser.isPremiumSubscriber
        }
        
        // Default: Allow LLaMA for all users (app is paid, not subscription-based)
        // In production, you might want to restrict this to premium users only
        return true
    }
}

enum DocumentError: LocalizedError {
    case contextNotAvailable
    case serviceNotConfigured
    case uploadFailed
    case encryptionFailed
    case vaultNotFound
    case contentBlocked(severity: ContentSeverity, categories: [ContentCategory], reason: String?)
    
    var errorDescription: String? {
        switch self {
        case .contextNotAvailable:
            return "Database context not available"
        case .serviceNotConfigured:
            return "Service not configured. Please ensure Supabase is properly initialized."
        case .uploadFailed:
            return "Failed to upload document"
        case .encryptionFailed:
            return "Failed to encrypt document"
        case .vaultNotFound:
            return "Vault not found"
        case .contentBlocked(let severity, let categories, let reason):
            var message = "Content blocked due to \(severity.rawValue) severity"
            if let reason = reason {
                message += ": \(reason)"
            }
            if !categories.isEmpty {
                message += " (Categories: \(categories.map { $0.rawValue }.joined(separator: ", ")))"
            }
            return message
        }
    }
}

File: platforms/android/app/src/main/java/com/khandoba/securedocs/service/DocumentService.kt
package com.khandoba.securedocs.service

import android.content.Context
import android.net.Uri
import android.util.Log
import com.khandoba.securedocs.config.AppConfig
import com.khandoba.securedocs.data.entity.DocumentEntity
import com.khandoba.securedocs.data.repository.DocumentRepository
import com.khandoba.securedocs.data.supabase.SupabaseService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.File
import java.io.FileInputStream
import java.util.Date
import java.util.UUID

class DocumentService(
    private val context: Context,
    private val documentRepository: DocumentRepository,
    private val supabaseService: SupabaseService,
    private val encryptionService: EncryptionService,
    private val documentIndexingService: DocumentIndexingService? = null
) {
    private val _documents = MutableStateFlow<List<DocumentEntity>>(emptyList())
    val documents: StateFlow<List<DocumentEntity>> = _documents.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _uploadProgress = MutableStateFlow(0.0)
    val uploadProgress: StateFlow<Double> = _uploadProgress.asStateFlow()
    
    suspend fun loadDocuments(vaultId: UUID) {
        _isLoading.value = true
        try {
            documentRepository.getDocumentsByVault(vaultId).collect { documentList ->
                _documents.value = documentList.filter { it.status == "active" && !it.isArchived }
            }
        } catch (e: Exception) {
            Log.e("DocumentService", "Error loading documents: ${e.message}")
        } finally {
            _isLoading.value = false
        }
    }
    
    suspend fun uploadDocument(
        vaultId: UUID,
        uri: Uri,
        name: String,
        uploadedByUserID: UUID
    ): Result<DocumentEntity> {
        _isLoading.value = true
        _uploadProgress.value = 0.0
        
        return try {
            // Read file
            val inputStream = context.contentResolver.openInputStream(uri)
                ?: return Result.failure(Exception("Cannot open file"))
            
            val fileData = inputStream.readBytes()
            inputStream.close()
            
            _uploadProgress.value = 0.3
            
            // Encrypt file
            val encryptedData = encryptionService.encrypt(fileData)
            val encryptionKey = encryptionService.generateKey()
            
            _uploadProgress.value = 0.6
            
            // Determine file type
            val mimeType = context.contentResolver.getType(uri) ?: "application/octet-stream"
            val fileExtension = name.substringAfterLast(".", "")
            val documentType = when {
                mimeType.startsWith("image/") -> "image"
                mimeType.startsWith("video/") -> "video"
                mimeType.startsWith("audio/") -> "audio"
                mimeType == "application/pdf" -> "pdf"
                mimeType.startsWith("text/") -> "text"
                else -> "other"
            }
            
            // Create initial document entity for indexing (before encryption)
            var documentName = name
            var aiTags = emptyList<String>()
            var extractedText: String? = null
            
            // Index document and generate intelligent name/tags (on unencrypted data)
            if (documentIndexingService != null && documentType == "image") {
                try {
                    // For images, we need to create a bitmap for OCR
                    val inputStream = context.contentResolver.openInputStream(uri)
                    val bitmap = android.graphics.BitmapFactory.decodeStream(inputStream)
                    inputStream?.close()
                    
                    if (bitmap != null) {
                        // Index document to extract text and generate tags/name
                        val indexedDoc = documentIndexingService.indexDocument(
                            document = DocumentEntity(
                                id = UUID.randomUUID(),
                                name = name,
                                fileExtension = fileExtension,
                                mimeType = mimeType,
                                documentType = documentType,
                                extractedText = null,
                                aiTags = emptyList()
                            ),
                            imageBitmap = bitmap
                        )
                        
                        documentName = indexedDoc.name
                        aiTags = indexedDoc.aiTags
                        extractedText = indexedDoc.extractedText
                        
                        bitmap.recycle()
                    }
                } catch (e: Exception) {
                    Log.e("DocumentService", "Indexing failed, using original name: ${e.message}")
                    // Continue with original name if indexing fails
                }
            }
            
            _uploadProgress.value = 0.5
            
            // Upload to Supabase Storage if enabled
            var storagePath: String? = null
            if (AppConfig.USE_SUPABASE) {
                try {
                    storagePath = "${vaultId}/${UUID.randomUUID()}.encrypted"
                    supabaseService.uploadFile(
                        bucket = AppConfig.ENCRYPTED_DOCUMENTS_BUCKET,
                        path = storagePath,
                        data = encryptedData
                    )
                    _uploadProgress.value = 0.8
                } catch (e: Exception) {
                    Log.e("DocumentService", "Failed to upload to Supabase Storage: ${e.message}")
                    // Continue with local storage
                }
            }
            
            // Create document entity with intelligent name and tags
            val document = DocumentEntity(
                id = UUID.randomUUID(),
                name = documentName, // Use intelligent name from indexing
                fileExtension = fileExtension,
                mimeType = mimeType,
                fileSize = fileData.size.toLong(),
                createdAt = Date(),
                uploadedAt = Date(),
                encryptedFileData = if (storagePath == null) encryptedData else null,
                encryptionKeyData = encryptionKey,
                isEncrypted = true,
                documentType = documentType,
                status = "active",
                vaultId = vaultId,
                uploadedByUserID = uploadedByUserID,
                aiTags = aiTags, // Use tags from indexing
                extractedText = extractedText // Store extracted text for search
            )
            
            // Save to database
            documentRepository.insertDocument(document)
            
            _uploadProgress.value = 1.0
            _isLoading.value = false
            
            Log.d("DocumentService", "✅ Document uploaded: ${document.name}")
            Result.success(document)
        } catch (e: Exception) {
            _isLoading.value = false
            _uploadProgress.value = 0.0
            Log.e("DocumentService", "Upload failed: ${e.message}")
            Result.failure(e)
        }
    }
    
    suspend fun downloadDocument(document: DocumentEntity): Result<ByteArray> {
        _isLoading.value = true
        return try {
            val encryptedData = if (document.encryptedFileData != null) {
                document.encryptedFileData!!
            } else if (AppConfig.USE_SUPABASE && document.id != null) {
                // Download from Supabase Storage
                val storagePath = "${document.vaultId}/${document.id}.encrypted"
                supabaseService.downloadFile(
                    bucket = AppConfig.ENCRYPTED_DOCUMENTS_BUCKET,
                    path = storagePath
                )
            } else {
                return Result.failure(Exception("Document data not available"))
            }
            
            // Decrypt
            val decryptedData = encryptionService.decrypt(encryptedData)
            
            _isLoading.value = false
            Result.success(decryptedData)
        } catch (e: Exception) {
            _isLoading.value = false
            Log.e("DocumentService", "Download failed: ${e.message}")
            Result.failure(e)
        }
    }
    
    suspend fun deleteDocument(document: DocumentEntity): Result<Unit> {
        _isLoading.value = true
        return try {
            // Delete from Supabase Storage if stored there
            if (AppConfig.USE_SUPABASE) {
                try {
                    val storagePath = "${document.vaultId}/${document.id}.encrypted"
                    supabaseService.deleteFile(
                        bucket = AppConfig.ENCRYPTED_DOCUMENTS_BUCKET,
                        path = storagePath
                    )
                } catch (e: Exception) {
                    Log.w("DocumentService", "Failed to delete from storage: ${e.message}")
                }
            }
            
            // Update status to deleted
            val updatedDocument = document.copy(status = "deleted")
            documentRepository.updateDocument(updatedDocument)
            
            _isLoading.value = false
            Result.success(Unit)
        } catch (e: Exception) {
            _isLoading.value = false
            Result.failure(e)
        }
    }
    
    suspend fun archiveDocument(document: DocumentEntity): Result<Unit> {
        return try {
            val updatedDocument = document.copy(isArchived = true)
            documentRepository.updateDocument(updatedDocument)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun bulkDeleteDocuments(documentIds: List<UUID>): Result<Unit> {
        _isLoading.value = true
        return try {
            val documents = _documents.value.filter { it.id in documentIds }
            documents.forEach { document ->
                deleteDocument(document)
            }
            _isLoading.value = false
            Result.success(Unit)
        } catch (e: Exception) {
            _isLoading.value = false
            Result.failure(e)
        }
    }
    
    suspend fun bulkArchiveDocuments(documentIds: List<UUID>): Result<Unit> {
        _isLoading.value = true
        return try {
            val documents = _documents.value.filter { it.id in documentIds }
            documents.forEach { document ->
                archiveDocument(document)
            }
            _isLoading.value = false
            Result.success(Unit)
        } catch (e: Exception) {
            _isLoading.value = false
            Result.failure(e)
        }
    }
}
